File: delivery/src/main/java/com/sparta/delivery/libs/RoleValidation.java
package com.sparta.delivery.libs;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Constraint(validatedBy = RoleValidator.class)
@Target({ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
public @interface RoleValidation {

    String[] roles() default {};

    String message() default "Invalid role";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};
}

File: hub/src/main/kotlin/com/sparta/hub/libs/RoleValidation.kt
package com.sparta.hub.libs

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass

@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.FUNCTION)
@Constraint(validatedBy = [RoleValidator::class])
annotation class RoleValidation(
    vararg val roles: String,
    val message: String = "Invalid role",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = []
)
